File: pythonStuff/leetcode378.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        
        def mergeSort(inputList):
            if len(inputList) < 2:
                return inputList
            leftArray = mergeSort(inputList[(len(inputList)//2):])
            rightArray = mergeSort(inputList[:(len(inputList)//2)])
            return merge(leftArray, rightArray)

        def merge(leftArray, rightArray):
            leftPointer, rightPointer = 0, 0
            combinedArr = []
            while leftPointer < len(leftArray) and rightPointer < len(rightArray):
                if leftArray[leftPointer] < rightArray[rightPointer]:
                    combinedArr.append(leftArray[leftPointer])
                    leftPointer += 1
                else:
                    combinedArr.append(rightArray[rightPointer])
                    rightPointer += 1
            combinedArr.extend(leftArray[leftPointer:])
            combinedArr.extend(rightArray[rightPointer:])
            return combinedArr
        
        fullList = []
        for x in range(len(matrix)):
            for y in range(len(matrix[x])):
                fullList.append(matrix[x][y])
        return mergeSort(fullList)[k-1]

File: javaStuff/leetcode378.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        int[] fullList = new int[(matrix.length*matrix[0].length)];
        int counter = 0;
        for(int x = 0; x < matrix.length; x++){
            for(int y = 0; y < matrix[x].length; y++){
                fullList[counter] = matrix[x][y];
                counter++;
            }
        }
        mergeSort(fullList);
        return fullList[k-1];
    }
    
    private static void mergeSort(int[] inputArray){
        int arrayLength = inputArray.length;
        if (arrayLength < 2){
            return;
        }
        int arrayMid = arrayLength / 2;
        int[] leftArray = new int[arrayMid];
        int[] rightArray = new int[arrayLength-arrayMid];
        for (int i = 0; i < arrayMid; i++){
            leftArray[i] = inputArray[i];
        }
        for (int i = arrayMid; i < arrayLength; i++){
            rightArray[i-arrayMid] = inputArray[i];
        }
        mergeSort(leftArray);
        mergeSort(rightArray);
        merge(inputArray, leftArray, rightArray);
    }
    
    private static void merge(int[] inputArray, int[] leftArray, int[] rightArray){
        int leftSize = leftArray.length;
        int rightSize = rightArray.length;
        
        int leftPointer = 0, rightPointer = 0, counter =0;
        while (leftPointer < leftSize && rightPointer < rightSize){
            if (leftArray[leftPointer] <= rightArray[rightPointer]){
                inputArray[counter] = leftArray[leftPointer];
                leftPointer++;
            }
            else{
                inputArray[counter] = rightArray[rightPointer];
                rightPointer++;
            }
            counter++;
        }
        while (leftPointer < leftSize){
            inputArray[counter] = leftArray[leftPointer];
            leftPointer++;
            counter++;
        }
        while (rightPointer < rightSize){
            inputArray[counter] = rightArray[rightPointer];
            rightPointer++;
            counter++;
        }
    }
}